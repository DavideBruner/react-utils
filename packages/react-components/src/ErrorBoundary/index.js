import PropTypes from 'prop-types';
import React from 'react';
import cn from 'classnames';

import WideError from '../WideError';

/** A wrapping component around the application.
 * This is used to display errors thrown anywhere in the component tree. */
export default class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  componentDidCatch(error, info) {
    const { onError } = this.props;
    this.setState({ hasError: true });
    onError(error, info);
  }

  render() {
    const { children, className } = this.props;
    const { hasError } = this.state;
    const classes = cn('rau-error-boundary', className);

    if (!hasError) {
      return children;
    }

    return (
      <div className={classes}>
        <WideError title="Bear with us">
          <p>Something went wrong in our end, please contact support@usabilla.com</p>
        </WideError>
      </div>
    );
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  /** A function to run when an error is thrown.
   * The function will be invoked with 2 arguments, the error message, and extra information */
  onError: PropTypes.func,
};

ErrorBoundary.defaultProps = {
  className: undefined,
  onError: undefined,
};
