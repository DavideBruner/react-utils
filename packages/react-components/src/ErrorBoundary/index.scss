.#{$rau-ns}error-boundary {
  height: 100vh;
}
