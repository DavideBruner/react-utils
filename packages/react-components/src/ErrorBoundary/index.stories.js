import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import React, { useState } from 'react';

import ErrorBoundary from './index';

import '../../index.scss';

function ErrorButtonDispatcher() {
  const [hasError, setError] = useState(false);

  if (hasError) {
    throw new ErrorBoundary('an unexpected error');
  }

  return (
    <button type="button" onClick={() => setError(true)}>
      Click to dispatch an error
    </button>
  );
}

storiesOf('ErrorBoundary', module)
  .add('When an unexpected error occured', () => (
    <ErrorBoundary onError={action('logger')}>
      <ErrorButtonDispatcher />
    </ErrorBoundary>
  ));
