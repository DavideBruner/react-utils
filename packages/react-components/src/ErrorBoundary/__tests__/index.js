import { render } from '@testing-library/react';
import React from 'react';

import ErrorBoundary from '../index';

describe('ErrorBoundary', () => {
  const handleErrorMock = jest.fn();

  beforeEach(() => {
    // Needed to remove the annoying error message from the throw below
    jest.spyOn(window.console, 'error').mockImplementation(() => false);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const { container } = render(
      <ErrorBoundary onError={handleErrorMock}>
        No error expected
      </ErrorBoundary>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders an error message when an error is thrown, and log the error with user data', () => {
    const error = new Error('foo');
    const ErrorChild = () => {
      throw error;
    };

    const { container } = render(
      <ErrorBoundary onError={handleErrorMock}>
        <ErrorChild />
      </ErrorBoundary>,
    );

    expect(container).toMatchSnapshot();
    expect(handleErrorMock).toHaveBeenCalledTimes(1);
    expect(handleErrorMock.mock.calls[0][0]).toBe(error);
  });
});
