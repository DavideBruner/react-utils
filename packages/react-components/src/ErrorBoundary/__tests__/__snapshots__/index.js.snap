// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders an error message when an error is thrown, and log the error with user data 1`] = `
<div>
  <div
    class="rau-error-boundary"
  >
    <div
      class="sb-flex sb-flex-row sb-align-items-center rau-wide-error"
    >
      <div
        class="sb-flex-column sb-col-4 sb-offset-2 rau-wide-error-message"
      >
        <h2
          class="sb-title--l sb-padding-bottom"
        >
          Bear with us
        </h2>
        <p>
          Something went wrong in our end, please contact support@usabilla.com
        </p>
      </div>
      <svg
        aria-label="Error illustration"
        height="434"
        role="img"
        width="193"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          
        
          <path
            d="M95.775 341.575c-2.171-2.725 2.724-27.19 3.59-31.16.577-2.647 3.174-13.453 7.79-32.419-.749 18.588-1.187 30.634-1.315 36.138-.191 8.257-.992 12.69 0 20.463.31 2.43.834 17.623-.43 15.862-1.935-2.696-8.45-7.396-9.635-8.884zm7.275 35.892c-.858 7.778-1.35 11.866-1.476 12.264-.598 1.892-2.209 5.623-.598 6.605 1.61.983 11.44 2.664 12.442 2.664 1.002 0 1.752-3.63 1.752-5.203 0-.365.288-2.129.863-5.292 3.713-9.743 5.999-16.897 6.858-21.462 2.329-12.368 1.643-26.659 1.14-32.446-.337-3.858 1.25-13.21 4.757-28.057 6.727-19.582 10.397-31.584 11.01-36.004.612-4.42-.168-9.6-2.341-15.536l-60.833 4.929-1.574 3.252v9.059c0 3.24-5.243 18.066-1.98 67.558 0 3.5.906 6.155 2.72 7.965 9.246 11.747 16.346 19.657 21.3 23.727 3.393 2.788 5.38 4.78 5.96 5.977zM153.928 194.336c5.202-9.014 8.865-15.235 10.99-18.66.47-.78.937-1.27 1.457-1.51.607-.28 1.218-.208 1.943.142.394.19 1.965 1.167 1.908 1.134.672.395 1.276.686 1.917.9l.047-.076c.323-.52.512-.716.975-.592 1.554.414 4.2 1.036 7.66 1.81l.871.193a987.27 987.27 0 0 0 9.672 2.095l.036-.139c.096-.403.126-.821.063-1.23-.18-1.162-1.068-2.074-3.01-2.593-3.287-.88-9.521-3.758-18.708-8.627-3.328-1.47-6.177-.573-8.68 2.77-.803 1.07-5.648 7.587-7.078 9.5l-.268.359c-3.771 5.039-6.409 8.479-7.935 10.305l.012.006 1.71.887 6.418 3.326zm18.128-17.858a.496.496 0 0 0 .013-.02zm.875.168a.433.433 0 0 0 .043.006zM50.189 206.021c-10.542-17.042-16.898-27.27-19.118-30.75a2.973 2.973 0 0 0-.683-.588c-.42-.25-.848-.313-1.42-.245-.304.036-1.641.306-1.768.328-1.22.22-2.462.339-4.042.339-6.086 0-11.002-.436-13.408-1.274-1.166-.405-2.73-1.73-3.635-3.04-.42-.61-.635-1.135-.614-1.444.015-.204.155-.305.655-.305.41 0 .702-.184.984-.529.012-.014.393-.526.526-.671.43-.467 1-.75 1.968-.845 2.237-.221 7.709-.221 16.382.002l.14-.016c4.36-1.147 7.257-.256 8.845 2.65.627 1.147 2.05 3.033 4.458 5.971 1.381 1.686 7.033 8.426 7.31 8.759.42.504.818.984 1.202 1.448 5.532 6.697 8.817 11.117 10.146 13.934l.08.12c.514.575.274 1.29-.967 2.319-1.332 1.105-3.683 2.39-7.041 3.837z"
            stroke="#C8D2DA"
          />
          
        
          <path
            d="M76 151.136c2.955-7.917 5.005-15.551 6.152-22.9 1.72-11.024 2.715-20.18 8.355-25.54 3.76-3.572 7.1-5.107 10.02-4.603.317 0 .882.183 1.693.547 1.257 0 11.305.829 14.779 11.903 2.316 7.383 3.565 18.957 3.747 34.721-.7 5.494-.343 8.577 1.073 9.25 2.124 1.008-15.177 1.639-13.269-10.854.559-3.659 3.294-2.65 5.344-12.09 3.105-1.427 4.92-4.799 4.473-9.788-.124-1.386-1.09-1.711-2.9-.976-.384-11.054-4.64-16.58-12.767-16.58-12.193 0-14.256 23.405-14.256 25.232 0 1.827-.235 5.68 3.79 14.202-2.684 7.543-8.095 10.035-16.234 7.476z"
            fill="#C8D2DA"
          />
          
        
          <path
            d="M81.575 152.728c7.434 9.345 13.432 14.017 17.997 14.017 4.537 0 9.691-3.752 15.464-11.255a2 2 0 0 1 1.66-.78c10.79.403 16.929 3.72 18.416 9.954 1.534 6.433 3.293 16.048 5.279 28.845l5.551-5.803a1 1 0 0 1 1.33-.103l7.45 5.69a1 1 0 0 1 .206 1.378c-6.998 9.765-11.566 15.406-13.705 16.923-3.292 2.335-7.046 1.774-10.227-2.588s-4.532-11.79-5.569-11.463c-1.038.328-.212 10.188 1.42 12.775 1.631 2.588 16.18 42.589 12.223 46.385-28.61.714-49.68 2.83-63.208 6.347a1 1 0 0 1-1.244-.847c-1.48-12.115-1.008-26.4 1.416-42.857.67-3.914-1.27-18.458-2.584-18.458-.856 0-6.108 16.345-11.032 16.859-3.21.335-7.993-4.475-14.347-14.429a1 1 0 0 1 .146-1.254l6.463-6.293a1 1 0 0 1 1.27-.104l2.544 1.776c.805-17.33 2.113-29.097 3.924-35.302 1.747-5.991 7.559-9.378 17.434-10.161a2 2 0 0 1 1.723.748z"
            fill="#C8D2DA"
            stroke="#C8D2DA"
          />
          
        
          <path
            d="M94.251 433.856c10.185 1.252 14.711-5.671 13.58-20.77-.45 0-1.405.733-2.543 0 0-1.321-.655-7.086-8.453-7.086-1.201 0-1.635 1.343-1.3 4.028l-4.46 10.59c-3.355 2.763-5.033 5.112-5.033 7.049-.42 3.46 2.316 5.523 8.21 6.19zM125.041 377.505c4.253.953 6.38 5.037 6.38 12.25-.496 3.098-3.303 4.854-8.421 5.268 1.372 4.483 2.058 6.916 2.058 7.301 0 .577.432 14.011 4.793 18.694 4.36 4.682 5.324 5.798 8.14 4.449 2.816-1.35 6.396-7.584 4.458-17.022-1.938-9.437-4.458-9.135-5.231-13.144-.774-4.008.459-14.254-.808-16.517-1.266-2.263-6.56-5.689-10.201-4.563-1.066.246-1.71.953-1.931 2.122a1 1 0 0 0 .763 1.162z"
            fill="#C8D2DA"
          />
          
        
          <path
            d="M152.111 70.241L141.444 81V70.241H132a3 3 0 0 1-3-3V32a3 3 0 0 1 3-3h58a3 3 0 0 1 3 3v35.241a3 3 0 0 1-3 3h-37.889z"
            fill="#00C97E"
          />
          
        
          <path
            d="M145 47h6v6h-6zM159 47h6v6h-6zM173 47h6v6h-6z"
            fill="#FFF"
          />
          
        
          <g>
            
          
            <path
              d="M55.818 61.302l-3.992-7.987 1.52-1.272C60.115 46.37 64 38.272 64 29.823 64 13.38 49.644 0 32 0 14.354 0 0 13.38 0 29.825c0 16.446 14.355 29.823 32 29.823 2.389 0 4.779-.25 7.095-.738l.718-.154 13.675 5.129c.71.266 1.542.059 2.047-.498a1.856 1.856 0 0 0 .283-2.085z"
              fill="#EC4A59"
            />
            
          
            <path
              d="M36.172 25.332c0-1.84-1.752-3.546-3.672-3.546-1.92 0-3.672 1.706-3.672 3.534.01.69-.36 1.327-.96 1.668a1.933 1.933 0 0 1-1.908 0c-.6-.34-.97-.978-.96-1.656C25 21.336 28.426 18 32.5 18c4.066 0 7.489 3.324 7.5 7.313v.014c.015 2.568-1.096 4.252-3.414 6.304-1.112.985-1.389 1.246-1.696 1.642-.35.452-.476.78-.476 1.254v.583c.01.69-.36 1.328-.96 1.669a1.933 1.933 0 0 1-1.908 0c-.6-.341-.97-.979-.96-1.657v-.595c0-1.31.464-2.516 1.27-3.556.394-.508.837-.965 1.435-1.506.174-.157.924-.82.773-.685 1.676-1.486 2.12-2.14 2.108-3.448zM32.5 39.129c1.058 0 1.915.866 1.915 1.935A1.926 1.926 0 0 1 32.5 43a1.926 1.926 0 0 1-1.915-1.936c0-1.069.857-1.935 1.915-1.935z"
              fill="#FFF"
              fill-rule="nonzero"
            />
            
        
          </g>
          
      
        </g>
      </svg>
    </div>
  </div>
</div>
`;

exports[`ErrorBoundary renders without crashing 1`] = `
<div>
  No error expected
</div>
`;
